### PR TITLE
chore(commit message): use commitplease to validate commit messages

### PR DESCRIPTION
**Pull request by alisianoi**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] ~~Tests for the changes have been added (for bug fixes / features)~~
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Other... Please describe: check that commit messages adhere to style guide
```

**What is the current behavior?** (You can also link to an open issue here)

Developers follow commit message style best they can

**What is the new behavior?**

An external npm module that checks commit messages for you

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

[commitplease](https://github.com/jzaefferer/commitplease) is an npm module that installs a git hook during `npm install` and checks that your commit message follows the style guide when you do `git commit`. If the message does not follow the style guide, the commit is aborted and an exact diagnostic is printed so that the developer can correct the commit message.

How to try out:

```
git checkout commitplease
mv .git/hooks/commit-msg .git/hooks/commit-msg.old
npm install
```

Now try to commit some changes with valid/invalid commit messages.

How to go back:

```
git checkout master
mv .git/hooks/commit-msg.old .git/hooks/commit-msg
```

Disclaimer: I am contributing to [commitplease](https://github.com/jzaefferer/commitplease) as part of the Google Summer of Code project for students.

There is also a twin pull request to angular/angular.js#14888

### CODE REVIEW

**Comment by vicb** [comment]
I often commit tmp commits during dev, would it be possible to 
1) only prevent pushes
2) have a way to force a push

**Comment by all3fox** [comment]
For temporary commits, you could either begin the commit message with:
 1. `WIP`, `wip` or `Wip` which usually stands for "Work in progress"
 2.  `fixup!` or `squash!` which is a way to tell `git` [to squash the commits later][1]

or you could do `git commit --no-verify` which does not trigger the git hook altogether (ie does not run the validation script at all)

[1]: https://robots.thoughtbot.com/autosquashing-git-commits

File: package.json [hunk 1]
```diff
@@ -115,5 +116,8 @@
     "webpack": "^1.12.6",
     "which": "~1",
     "yargs": "^3.31.0"
+  },
+  "commitplease": {
```
Comment (alexeagle) [inline package.json]: should we include types as well?
https://www.npmjs.com/package/commitplease#angularjs
they are listed here:
https://github.com/angular/angular/blob/master/CONTRIBUTING.md#type

Comment (all3fox) [inline package.json]: no, those types are actually default values inside commitplease and can be left out

But you could redefine them in `package.json` if needed, as shown in the your link

Comment (alexeagle) [inline package.json]: I usually use WIP at the beginning of a tmp commit message, so this seems
like a good idea to me. Any objection @vicb ?

**Comment by vicb** [comment]
Can we use "f!" fixup is too long ? 

I -1 to merge this before:
- It is documented in the CONTRIBUTOR or DEVELOPPER guide (don't remember the best fir on top of my head) - how to escape from the message dictature, ...
- The error message points to this doc.

I think it would be a really bad dev XP to have a cryptic error message on commits.

**Comment by vicb** [comment]
@all3fox sorry for the delay in processing.

I would be interested in merging this PR.

Is there a way to test the commits message on the CI ? (fails if the commit message do not follow our conventions)

Thanks

**Comment by all3fox** [comment]
Hello @vicb 

Currently, commitplease checks its `argv[0]` which can be ([here is the code][2]):

1. The `.git/COMMIT_EDITMSG` path supplied by the `commit-msg` hook. Then commitplease reads the contents of this file and validates them.

2. Something that can be used by git log to filter commits (like `master` to show that branch or `master..feature` to show new commits on branch feature, [more examples here][1])

So, for instance you could have a shell script that creates a git repository, installs commitlpease there, configures it however you like via `package.json` and then begins the cycle of adding commits and reading git responses.

Finally, on our Travis CI we use this [test.js][3] script that basically creates a bunch of `messageX` -- `profilesX` pairs of lists. Each such pair corresponds to a particular style of commits (`message0` and `profile0` are for jQuery, `message1` and `profile1` are for Angular, `message9` and `profile9` are for common parts). By default, a profile from `profileX` must successfully validate all the messages from `messageX`. Alternatively, for every message you could say that the profile must be in `rejects` or `reasons`, which means that the profile simply has to reject the message or reject it with *and* print a specific error message. And you could also cherry-pick a subset of tests, like with `jquery4` [profile here][4] (it is not in any `profilesX` list but is manually inserted in some places)

So, what I am saying is you could take the main test function from [line 802][5] and then roll your own Travis CI test script with `messageX` --- `profilesX` lists.

Tell me what you would prefer and I may help with that.

[1]: https://github.com/jzaefferer/commitplease/#global-install
[2]: https://github.com/jzaefferer/commitplease/blob/c9ab536eee428d4a72549886ae93ebd7fd7d6cd9/index.js#L78
[3]: https://github.com/jzaefferer/commitplease/blob/c9ab536eee428d4a72549886ae93ebd7fd7d6cd9/test.js
[4]: https://github.com/jzaefferer/commitplease/blob/c9ab536eee428d4a72549886ae93ebd7fd7d6cd9/test.js#L50
[5]: https://github.com/jzaefferer/commitplease/blob/c9ab536eee428d4a72549886ae93ebd7fd7d6cd9/test.js#L802

**Comment by jzaefferer** [comment]
You should also be able to call the commitplease executable in a local install on Travis as described here: https://github.com/jzaefferer/commitplease/blob/master/README.md#global-install

**Comment by alxhub** [comment]
Closing as outdated, please sync and re-open if needed.

**Pull request closed without merging**